runtime/domain.c

```diff
@@ -1070,6 +1076,38 @@ CAMLexport _Atomic caml_timing_hook caml_domain_terminated_hook =
 
 static void domain_terminate(void);
 
+static void domain_sync_result(value term_sync, value res_or_exn)
+{
+  CAMLparam1(term_sync);
+  CAMLlocal1(res);
+  if (Is_exception_result(res_or_exn)) {
+    res = Extract_exception(res_or_exn);
+    /* [Finished (Error res)] from stdlib/domain.ml */
+    res = caml_alloc_1(1, res);
+    res = caml_alloc_1(0, res);
+  } else {
+    res = res_or_exn;
```

Comment:
I would have expected an allocation of `Ok res_or_exn` here. Unless I am missing something, (1) the code is wrong and (2) our testsuite needs beefing up on `Domain.join`.

Reply:
Ouch, well spotted! Besides my obvious oversight, I'll look into why the testsuite did not catch this.

Reply:
You convinced me that I made a mistake but it turns out that it is correct. `Domain.spawn` already returns `Finished (Ok res)` on the OCaml side. The testsuite would have caught the error, as well.

I'll add a comment.

Reply:
Ah, indeed!

I wonder if we should make the code more regular by changing `spawn` to return the result directly. Changing it to return an `('a, exn) result` value would be very simple, but we could go further by returning an `'a` and raising the corresponding exceptions -- I think there would be opportunities for simplifying the code there.

```ocaml
      match
        DLS.create_dls ();
        DLS.set_initial_keys pk;
        let res = f () in
        res
      with
      | res ->
        do_at_exit ();
        res
      | exception exn ->
        let bt = Printexc.get_raw_backtrace () in
        (try do_at_exit () with _ -> ());
        Printexc.raise_with_backtrace bt exn
```

Reply:
You mean `body`, right?
```
Error: Wrong link order: Stdlib__Domain depends on Stdlib__Printexc
```

Reply:
> let bt = Printexc.get_raw_backtrace () in

I left this out as I could not fix the dependency issues. I'll let anyone looking to fix #12362 deal with this!